zproject/backends.py

```diff
@@ -1556,13 +1563,68 @@ def get_data_from_redis(cls, key: str) -> Optional[Dict[str, Any]]:
         if key.startswith('saml_token_'):
             # Safety if statement, to not allow someone to poke around arbitrary redis keys here.
             data = get_dict_from_redis(redis_client, "saml_token_{token}", key)
-        if data is None:
-            # TODO: We will need some sort of user-facing message
-            # about the authentication session having expired here.
-            logging.info("SAML authentication failed: bad RelayState token.")
 
         return data
 
+    @classmethod
+    def get_issuing_idp(cls, SAMLResponse: str) -> Optional[str]:
+        """
+        Given a SAMLResponse, returns which of the configured IdPs is declared as the issuer.
+        This value MUST NOT be trusted as the true issuer!
+        The signatures are not validated, so it can be tampered with by the user.
+        That's not a problem for this function,
+        and true validation happens later in the underlying libraries, but it's important
+        to note this detail. The purpose of this function is merely as a helper to figure out which
+        of the configured IdPs' information to use for parsing and validating the response.
+        """
+        try:
+            resp = OneLogin_Saml2_Response(settings={}, response=SAMLResponse)
+            issuers = resp.get_issuers()
+        except cls.SAMLRESPONSE_PARSING_EXCEPTIONS as e:
+            logging.info("Error while parsing SAMLResponse: %s: %s", e.__class__.__name__, e)
+            return None
+
+        for idp_name, idp_config in settings.SOCIAL_AUTH_SAML_ENABLED_IDPS.items():
+            if idp_config['entity_id'] in issuers:
+                return idp_name
+
+        return None
+
+    def get_relayed_params(self) -> Dict[str, Any]:
+        request_data = self.strategy.request_data()
+        if 'RelayState' not in request_data:
+            return {}
+
+        # Get the relevant params that we transported in the RelayState:
+        relay_state = request_data['RelayState']
+        try:
+            data = ujson.loads(relay_state)
+            # Right now we only support transporting subdomain through json in
+            # RelayState, but this format is nice in that it allows easy extensibility here.
+            relayed_params = {'subdomain': data.get('subdomain')}
+        except (ValueError, TypeError):
+            # Not JSON data. The other possibility is it being a token under which
+            # we stored associated data in the first step of authentication.
+            relayed_params = self.get_data_from_redis(relay_state) or {}
+
+        return relayed_params
```

Comment:
See my comment on the next function, but I think this would be better with `return` statements too.